File: planctl/deps.py
"""Dependency graph operations."""

from __future__ import annotations


def has_cycle(graph: dict, task_id: str, visited: set, rec_stack: set) -> list[str]:
    """DFS cycle detection. Returns cycle path or empty list."""
    visited.add(task_id)
    rec_stack.add(task_id)

    for dep in graph.get(task_id, {}).get("depends_on", []):
        if dep not in visited:
            cycle = has_cycle(graph, dep, visited, rec_stack)
            if cycle:
                return [task_id] + cycle
        elif dep in rec_stack:
            return [task_id, dep]

    rec_stack.remove(task_id)
    return []


def detect_cycles(graph: dict) -> list[str] | None:
    """Run has_cycle from each unvisited node. Returns cycle path or None."""
    visited: set[str] = set()
    for task_id in graph:
        if task_id not in visited:
            cycle = has_cycle(graph, task_id, visited, set())
            if cycle:
                return cycle
    return None


def find_dependents(task_id: str, all_tasks: dict) -> list[str]:
    """All tasks that directly or transitively depend on task_id."""
    dependents: set[str] = set()
    to_check = [task_id]
    checked: set[str] = set()

    while to_check:
        checking = to_check.pop(0)
        if checking in checked:
            continue
        checked.add(checking)

        for tid, tdata in all_tasks.items():
            if tid in checked or tid in dependents:
                continue
            deps = tdata.get("depends_on", [])
            if checking in deps:
                dependents.add(tid)
                to_check.append(tid)

    return list(dependents)


def validate_dependency(
    task_id: str, dep_id: str, epic_id: str, existing_tasks: set
) -> str | None:
    """Validate a dependency. Returns error message or None."""
    from planctl.ids import epic_id_from_task, is_task_id

    if not is_task_id(dep_id):
        return f"Invalid task ID format: {dep_id}"

    if task_id == dep_id:
        return f"Task cannot depend on itself: {dep_id}"

    try:
        dep_epic = epic_id_from_task(dep_id)
    except ValueError:
        return f"Invalid dependency ID: {dep_id}"

    if dep_epic != epic_id:
        return f"Dependency {dep_id} is in a different epic ({dep_epic}), must be in {epic_id}"

    if dep_id not in existing_tasks:
        return f"Dependency {dep_id} does not exist"

    return None

File: plugins/plan/src/deps.ts
// Dependency-graph operations — the byte-parity port of planctl/deps.py.
//
// detectCycles / findDependents retain the Python DFS path shape. detectCycles
// owns a stable lexicographic traversal so cycle strings do not depend on object
// insertion order, directory enumeration order, or the host platform.

/** A node's record in the dependency graph: only `depends_on` is consulted. */
export interface DepNode {
  depends_on?: string[];
}

export type DepGraph = Record<string, DepNode>;

/** DFS from `taskId`, returning a cycle path or [] when none is reachable.
 * Mirrors deps.has_cycle: a back-edge into the recursion stack returns
 * `[taskId, dep]`; the recursive arm prepends `taskId` onto the discovered
 * cycle. `visited` and `recStack` are shared across the whole detect pass. */
export function hasCycle(
  graph: DepGraph,
  taskId: string,
  visited: Set<string>,
  recStack: Set<string>,
): string[] {
  visited.add(taskId);
  recStack.add(taskId);

  for (const dep of [...(graph[taskId]?.depends_on ?? [])].sort()) {
    if (!visited.has(dep)) {
      const cycle = hasCycle(graph, dep, visited, recStack);
      if (cycle.length > 0) {
        return [taskId, ...cycle];
      }
    } else if (recStack.has(dep)) {
      return [taskId, dep];
    }
  }

  recStack.delete(taskId);
  return [];
}

/** Run hasCycle from each unvisited node, returning the first cycle path or
 * null when the graph is acyclic. Nodes and adjacency lists use one canonical
 * lexicographic order, making the surfaced cycle stable for every caller. */
export function detectCycles(graph: DepGraph): string[] | null {
  const visited = new Set<string>();
  for (const taskId of Object.keys(graph).sort()) {
    if (!visited.has(taskId)) {
      const cycle = hasCycle(graph, taskId, visited, new Set<string>());
      if (cycle.length > 0) {
        return cycle;
      }
    }
  }
  return null;
}

/** Every task that directly OR transitively depends on `taskId`. Mirrors
 * deps.find_dependents: a BFS over the reverse edges (`allTasks[*].depends_on`
 * containing the current node), returned in discovery order. The returned list
 * order follows Python's set-to-list conversion only loosely; callers that need
 * a stable order sort the result. */
export function findDependents(taskId: string, allTasks: DepGraph): string[] {
  const dependents = new Set<string>();
  const toCheck: string[] = [taskId];
  const checked = new Set<string>();

  while (toCheck.length > 0) {
    const checking = toCheck.shift() as string;
    if (checked.has(checking)) {
      continue;
    }
    checked.add(checking);

    for (const [tid, tdata] of Object.entries(allTasks)) {
      if (checked.has(tid) || dependents.has(tid)) {
        continue;
      }
      const deps = tdata.depends_on ?? [];
      if (deps.includes(checking)) {
        dependents.add(tid);
        toCheck.push(tid);
      }
    }
  }

  return [...dependents];
}
